contest.hpp

```diff
@@ -185,23 +206,33 @@ export class Contest {
 
     void concede(Player player)
     {
-        if (status != Status::ON_GOING)
+        if (status != Status::ON_GOING){
+            logger->error("status:{}",(int)status);
+            logger->critical("Concede: Contest not started");
             throw std::logic_error("Contest not started");
-        if (players[current.role] != player)
+        }
+        if (players[current.role] != player){
+            logger->error("In {} turn",players[current.role].role.map("black","white","none"));
+            logger->critical("Concede: " + player.name + " not allowed to concede");
             throw std::logic_error(player.name + " not allowed to concede");
-
+        }
         status = Status::GAME_OVER;
         win_type = WinType::GIVEUP;
         winner = -player.role;
     }
 
     void overtime(Player player)
     {
-        if (status != Status::ON_GOING)
+        if (status != Status::ON_GOING){
+            logger->error("status:{}",(int)status);
+            logger->critical("Overtime: Contest not started");
             throw std::logic_error("Contest not started");
-        if (players[current.role] != player)
+        }
+        if (players[current.role] != player){
+            logger->error("In {}'s turn",current.role.map("black","white","none"));
+            logger->critical("Overtime: not in " + player.name + "'s turn");
```

Comment:
Is logging before throwing an exception a recommended way? Or logging when catching them?

Reply:
Logging when catching them is a good idea.

Reply:
But catching them need revise too much, logging befor throwing an exception could make it clearer, and we could change critical to error if we could catch and solve them.